pkg/secrets/annotations/annotation_parser.go

```diff
@@ -15,7 +16,14 @@ import (
 // by Kubernetes via the Downward API, based on Pod annotations that are defined
 // in a deployment manifest.
 func FromFile(path string) (map[string]string, error) {
-	annotationsFile, err := os.OpenFile(path, os.O_RDONLY, os.ModePerm)
+	// Use the standard OS
+	return fromFile(vfs.OS(), path)
+}
+
+// fromFile performs the work of FromFile(), and provides a function
+// entrypoint that allows filesystem mocking for test purposes.
+func fromFile(fs vfs.Filesystem, path string) (map[string]string, error) {
```

Comment:
I like the general approach, and happy to approve but I wanted to run by you a suggestion in the spirit of avoiding adding a dependency. The suggestion works because the function that actually does the parsing, `ParseReader`, just needs a reader. Files just need closing, so the lowest common denominator in terms of interfaces is the `io.ReadCloser`.

```
package main

import (
	"fmt"
	"io"
	"io/ioutil"
	"os"
	"strings"
)

// fileOpener is the interface that captures the dependency injection, much like vfs.Filesystem in your example
type fileOpener func(name string, flag int, perm os.FileMode) (io.ReadCloser, error)

// some utilities for mocking
func mockReadCloser(contents string) io.ReadCloser {
	return ioutil.NopCloser(strings.NewReader(contents))
}
func mockFileOpenerGenerator(store map[string]io.ReadCloser) fileOpener {
	return func(name string, flag int, perm os.FileMode) (io.ReadCloser, error) {
		rc, ok := store[name]
		if ok {
			return rc, nil
		}
		return nil, fmt.Errorf("file not found")
	}
}

func main() {
	// mock fileOpener
	fo := mockFileOpenerGenerator(map[string]io.ReadCloser{
		"file": mockReadCloser("file contents"),
	})

	// real fileOpener
	// fo := func(name string, flag int, perm os.FileMode) (io.ReadCloser, error) {
	//	return os.OpenFile(name, flag, perm)
	// }

	rc, err := fo("file", os.O_RDONLY, os.ModePerm)
	if err != nil {
		panic(err)
	}
	defer rc.Close()

	out, err := ioutil.ReadAll(rc)
	fmt.Println(string(out), err)
}
```

Reply:
Yeah, I think it's worth the effort to avoid a new dependency.